### PR TITLE
add configurable pause annotation support

### PR DESCRIPTION
**Pull request by xh3b4sd**
## Checklist

- [x] Update changelog in CHANGELOG.md.
- [ ] Update roadmap in ROADMAP.md.

## Context 

Resolves https://github.com/giantswarm/giantswarm/issues/10950.

### CODE REVIEW

**Comment by xh3b4sd** [comment]
I still want to write an integration test to make sure pausing works, but you can have a look at the code already.

File: controller/controller.go [hunk 1]
```diff
@@ -554,6 +599,16 @@ func ProcessUpdate(ctx context.Context, obj interface{}, resources []resource.In
...
 }
 
+func hasAnnotation(m map[string]string, k string, v string) bool {
+	for a, b := range m {
+		if a == k && b == v {
+			return true
+		}
+	}
+
+	return false
```
Comment (kopiczko) [inline controller/controller.go:609]:
```suggestion
	return m[k] == v
```

File: controller/controller.go [hunk 2]
```diff
@@ -53,6 +58,13 @@ const (
 	loggerKeyVersion    = "version"
 )
 
+var (
+	defaultPauseAnnotations = map[string]string{
+		"cluster.x-k8s.io/paused":             "true",
+		"operatorkit.giantswarm.io/reconcile": "false",
```
Comment (kopiczko) [inline controller/controller.go:64]: How about this?

```suggestion
		"operatorkit.giantswarm.io/paused": "true",
```

Comment (xh3b4sd) [reply controller/controller.go:64]: Not a fan but I see that the consistency is striking.

**Comment by xh3b4sd** [comment]
I added the integration test and wrote some docs. Please have another look @kopiczko and @giantswarm/team-firecracker-engineers.

**Comment by xh3b4sd** [comment]
The logs of the `pause` integration test show that the reconciliation is halted. 

```
{"caller":"github.com/giantswarm/k8sclient/v3/pkg/k8scrdclient/crd_client.go:89","level":"debug","message":"creating CRD `drainerconfigs.core.giantswarm.io`","time":"2020-08-10T09:51:59.21932+00:00"}
{"caller":"github.com/giantswarm/k8sclient/v3/pkg/k8scrdclient/crd_client.go:100","level":"debug","message":"created CRD `drainerconfigs.core.giantswarm.io`","time":"2020-08-10T09:51:59.235645+00:00"}
{"caller":"github.com/giantswarm/k8sclient/v3/pkg/k8scrdclient/crd_client.go:124","level":"debug","message":"updating CRD `drainerconfigs.core.giantswarm.io`","time":"2020-08-10T09:51:59.245239+00:00"}
{"caller":"github.com/giantswarm/exporterkit/collector/set.go:48","level":"debug","message":"booting collector","time":"2020-08-10T09:52:00.056347+00:00"}
{"caller":"github.com/giantswarm/k8sclient/v3/pkg/k8scrdclient/crd_client.go:124","level":"debug","message":"updating CRD `drainerconfigs.core.giantswarm.io`","time":"2020-08-10T09:52:00.291809+00:00"}
{"caller":"github.com/giantswarm/k8sclient/v3/pkg/k8scrdclient/crd_client.go:137","level":"debug","message":"updated CRD `drainerconfigs.core.giantswarm.io`","time":"2020-08-10T09:52:00.3301+00:00"}
{"caller":"github.com/giantswarm/operatorkit/controller/controller.go:455","controller":"test-controller","level":"debug","loop":"4","message":"found pause annotation","time":"2020-08-10T09:52:06.357522+00:00"}
{"caller":"github.com/giantswarm/operatorkit/controller/controller.go:456","controller":"test-controller","level":"debug","loop":"4","message":"cancelling reconciliation","time":"2020-08-10T09:52:06.357573+00:00"}
{"caller":"github.com/giantswarm/operatorkit/controller/controller.go:455","controller":"test-controller","level":"debug","loop":"5","message":"found pause annotation","time":"2020-08-10T09:52:08.351224+00:00"}
{"caller":"github.com/giantswarm/operatorkit/controller/controller.go:456","controller":"test-controller","level":"debug","loop":"5","message":"cancelling reconciliation","time":"2020-08-10T09:52:08.351284+00:00"}
{"caller":"github.com/giantswarm/operatorkit/controller/controller.go:455","controller":"test-controller","level":"debug","loop":"6","message":"found pause annotation","time":"2020-08-10T09:52:10.392658+00:00"}
{"caller":"github.com/giantswarm/operatorkit/controller/controller.go:456","controller":"test-controller","level":"debug","loop":"6","message":"cancelling reconciliation","time":"2020-08-10T09:52:10.392717+00:00"}
{"caller":"github.com/giantswarm/operatorkit/controller/controller.go:455","controller":"test-controller","level":"debug","loop":"7","message":"found pause annotation","time":"2020-08-10T09:52:12.435545+00:00"}
{"caller":"github.com/giantswarm/operatorkit/controller/controller.go:456","controller":"test-controller","level":"debug","loop":"7","message":"cancelling reconciliation","time":"2020-08-10T09:52:12.435602+00:00"}
```

See https://app.circleci.com/pipelines/github/giantswarm/operatorkit/313/workflows/534d7f8b-40ce-4710-882f-b3570227d68e/jobs/8512.

**Comment by xh3b4sd** [comment]
Going with that. Let me know if something is off.

File: controller/controller.go [hunk 3]
```diff
@@ -53,6 +58,13 @@ const (
 	loggerKeyVersion    = "version"
 )
 
+var (
+	defaultPauseAnnotations = map[string]string{
+		"cluster.x-k8s.io/paused":          "true",
+		"operatorkit.giantswarm.io/paused": "true",
```
Comment (fiunchinho) [inline controller/controller.go:64]: What would be the use case that requires a different annotation than the one used upstream?

Comment (xh3b4sd) [reply controller/controller.go:64]: `operatorkit` is an abstract framework that can be used for all kinds of operators. CAPI is a niche within the Kubernetes ecosystem. When you build an nginx operator why would you want to have to know anything about Cluster API? I mean, I would not want to. I want to build my nginx operator and maybe use the pausing annotation for my field of expertise. 🤗

**Pull request merged**